parlai/agents/bart/bart.py

```diff
@@ -0,0 +1,181 @@
+        """
+        Override to add init-fairseq-model arg.
+        """
+        TransformerGeneratorAgent.add_cmdline_args(argparser)
+        group = argparser.add_argument_group('Bart Args')
+        group.add_argument(
+            '--init-fairseq-model',
+            type=str,
+            default=None,
+            help='fairseq checkpoint for bart',
+        )
+
+    def __init__(self, opt: Opt, shared: TShared = None):
+        if not shared:
+            opt = self._initialize_bart(opt)
+        super().__init__(opt, shared)
+
+    def _initialize_bart(self, opt: Opt) -> Opt:
+        """
+        Download and convert BART pre-trained models.
+
+        Additionally, convert `init-fairseq-model` if necessary.
+
+        :param opt:
+            ParlAI-parsed options
+
+        :return opt:
+            return opt with BART-specific args.
+        """
+        if not opt.get('converting'):
+            download(opt['datapath'])
+            opt['init_model'] = os.path.join(
+                opt['datapath'], 'models/bart/bart_large/model'
+            )
+        if opt.get('init_fairseq_model'):
+            opt = self._convert_model(opt)
+        opt.update(BART_ARGS)
+        compare_init_model_opts(opt, opt)
+        return opt
+
+    def _convert_model(self, opt: Opt) -> Opt:
+        """
+        Convert fairseq init model to ParlAI Model.
+
+        :param opt:
+            options
+
+        :return opt:
+            return opt with new init_model path
+        """
+        model_name = os.path.split(opt['init_fairseq_model'])[-1]
+        args = CONVERSION_ARGS.copy()
+
+        args['input'] = [opt['init_fairseq_model']]
+        if opt.get('model_file') and not os.path.exists(opt['model_file']):
+            args['output'] = opt['model_file']
+        else:
+            args['output'] = os.path.join(
+                opt['datapath'], 'models/converted_fairseq_models/', model_name
+            )
+
+        ConversionScript.main(**args)
+        opt['init_model'] = args['output']
+        return opt
+
+    def build_model(self) -> BartModel:
+        """
+        Build and return model.
+        """
+        model = BartModel(self.opt, self.dict)
+        if self.opt['embedding_type'] != 'random':
+            self._copy_embeddings(
+                model.encoder.embeddings.weight, self.opt['embedding_type']
+            )
+        return model
+
+    def _set_text_vec(
+        self, obs: Message, history: History, truncate: Optional[int]
+    ) -> Message:
+        """
+        Override to prepend start token and append end token.
+        """
+        obs = super()._set_text_vec(obs, history, truncate)
+        if 'text' not in obs or 'text_vec' not in obs:
+            return obs
+        vec = obs['text_vec']
+        if truncate is not None:
+            vec = torch.LongTensor(
+                self._check_truncate(obs['text_vec'], truncate - 2, True)
+            )
+        obs.force_set(
```

Comment:
can we call `self.add_start_end_tokens` here instead?

Reply:
will do